src/rpc/mining.cpp

```diff
@@ -572,24 +580,36 @@ UniValue getblocktemplate(const UniValue& params, bool fHelp)
             nTransactionsUpdatedLastLP = nTransactionsUpdatedLast;
         }
 
-        // Release the wallet and main lock while waiting
-        LEAVE_CRITICAL_SECTION(cs_main);
         {
-            checktxtime = boost::get_system_time() + boost::posix_time::minutes(1);
+            checktxtime = boost::get_system_time() + boost::posix_time::seconds(10);
 
             boost::unique_lock<boost::mutex> lock(csBestBlock);
             while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
             {
-                if (!cvBlockChange.timed_wait(lock, checktxtime))
-                {
-                    // Timeout: Check transactions for update
-                    if (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP)
-                        break;
-                    checktxtime += boost::posix_time::seconds(10);
+                // Before waiting, generate the coinbase for the block following the next
+                // block (since this is cpu-intensive), so that when next block arrives,
+                // we can quickly respond with the template for following block.
+                if (!have_next_coinbase_tx && IsShieldedMinerAddress(minerAddress)) {
+                    CoinbaseTransaction(Params(), next_coinbase_tx, CAmount{0}, minerAddress, chainActive.Tip()->nHeight + 2);
```

Comment:
I think it would be nicer if this returned the newly constructed transaction rather than mutating one in place. Is this kind of initialize-to-invalid-state-then-mutate really the preferred c++ idiom?